### DISCUSSION TITLE
Ability to tune the final webpack config

### DISCUSSION THREAD

**Discussion by kinetifex** [vercel/next.js#27587] [Ideas]
### Describe the feature you'd like to request

For some advanced cases in our apps, we need to modify the final webpack config generated by Next internals. Specifically - we need the ability to tune up the rules and entries modified by Next which are added _after_ the existing `webpack` config method.

### Describe the solution you'd like

A `webpackFinal` configuration method. See proposed implementation: https://github.com/vercel/next.js/pull/27544

The name draws from [Storybook's webpack config method](https://storybook.js.org/docs/react/configure/webpack#extending-storybooks-webpack-config).

This will allow webpack modification without interfering with Next.js internals. It will allow teams to fine-tune and more readily experiment with various webpack configurations in their Next.js app.

### Describe alternatives you've considered

The existing `webpack` config method could move to be invoked later - but this would be breaking for existing uses. We could suggest/request changes to how rules and entries are set by Next.js - but that is restricting.

**Comment by timneutkens** [comment]
This does not explain the exact use-case that you're trying to solve by introducing such a feature, can you please add that to the feature request?

**Comment by kinetifex** [comment]
Hey @timneutkens - we have an exact use-case now, which I'll describe below, but this has come up before for us that I find it a more general need.

> we need the ability to tune up the rules and entries modified by Next which are added after the existing webpack config method.

I'll provide more details specifically around entries, but the premise is the same for rules.

**Premise**: We have some heavy content pages that only need runtime JS in a few places and we are trying to cut down on the amount of JS we ship to the browser. We have found one solution by having custom webpack entries for pages to set up partial hydration of components on the page. By utilizing `unstable_runtimeJS` in a custom `_document.js`, we can fine-tune what scripts are loaded and use alternative entry scripts for pages.

**Problem**: In the Next.js custom webpack config method, `entry` is an async function. The results of this function are what get passed along to webpack as the final config to Webpack and there is no opportunity to adjust the end results.

At first, we were wrapping `webpack.entry` in our next config. However, the internal behavior of the entry function also changed in a patch version of Next.js (v10.2.0 -> v10.2.1 per https://github.com/vercel/next.js/pull/24656) - so this approach is unreliable. For v10.2.1+, to add custom entries as desired, we are required to monkey-patch Next.js internals to get around what `updatedEntry` is doing to our entries, which occurs _after_ any customized entries are in place.

**Proposal**: A solution to this problem would be to allow the webpack config to be modified by users after Next.js has done its work. Besides solving this one particular account, having an avenue for developers to make adjustments after Next.js has done its configuration would be helpful for experimentation. So, if not a new `webpackFinal` method, then moving `webpack` to execute _after_ Next.js internals is desirable.

**Reply by kinetifex** [reply]
I put together this demo app to showcase what we're experimenting with so far regarding custom entries.

Please take a look:
https://github.com/agerard-godaddy/next-custom-entries-demo

In particular, in order to manage the webpack config as needed, this app patches a `webpackFinal` config method [on postinstall](https://github.com/agerard-godaddy/next-custom-entries-demo/blob/main/package.json#L10) and uses the method in a [next.config plugin](https://github.com/agerard-godaddy/next-custom-entries-demo/blob/main/custom-entries/plugin.js#L47).

**Reply by agerard-godaddy** [reply]
Hey @timneutkens - did you get a chance to look this over? Does my explanation help describe the motive for the ask?
